src/draw_target.rs

```diff
@@ -498,8 +498,21 @@ impl DrawState {
             term.move_cursor_up(n.saturating_sub(1))?;
         }
 
-        let width = term.width() as usize;
-        let visual_lines = self.visual_line_count(.., width);
+        let term_width = term.width() as usize;
+
+        // The number of text lines that are contained in this draw state
+        let text_line_count = self.orphan_lines_count;
+
+        // Here we calculate the terminal vertical height each of those groups require when
+        // printing, taking wrapping into account.
+        let orphan_visual_line_count = self.visual_line_count(..text_line_count, term_width);
+        let bar_height = self.visual_line_count(text_line_count.., term_width);
+        let visual_lines = self.visual_line_count(.., term_width);
+
+        // Sanity checks
+        debug_assert!(visual_lines == orphan_visual_line_count + bar_height);
+        debug_assert!(self.orphan_lines_count <= self.lines.len());
```

Comment:
We need `visual_lines` for the `shift` definition, but I'd prefer to defer the calculation of `orphan_visual_line_count` until after the `shift` definition.